### PR TITLE
[ioBroker] Add smartName to StateObject definition, add EnumObject definition, fix typo

### LINKED ISSUE #251: Missing smartName in TypeScript definition
The iot adapter (and possibly others) uses the `common.smartName` property to know how the state should be named e.g. in Alexa.

`smartName` can be `'ignore'` (special case: the state is to be ignored), a simple string or an object containing at least:
```
{
  byOn: boolean;
  en: string;
}
```
It is very much possible that there are more properties in the object.
You can find more infos here:
https://github.com/ioBroker/ioBroker.iot/blob/master/lib/alexaSmartHomeV2.js

### LINKED ISSUE #252: Error in PartialOtherObject.common TypeScript definition
```
        interface OtherObject extends BaseObject {
            type: 'adapter' | 'config' | 'enum' | 'group' | 'host' | 'info' | 'instance' | 'meta' | 'script' | 'user';
            common: OtherCommon;
        }
        interface PartialOtherObject extends Partial<Omit<OtherObject, 'common'>> {
            common?: Partial<ObjectCommon>;
        }
```
The `common` of a PartialOtherObjects should be of type `Partial<OtherCommon>` and not `Partial<ObjectCommon>`.

### LINKED ISSUE #253: Missing TypeScript definition for EnumObject
In the TypeScript definition, enum objects are currently handled as `OtherObject` but enums require (at least) the `members` property, thus it would be good to have a definition for EnumObject (otherwise the compiler complains about the unknown `members` property.

### PR DESCRIPTION
**Pull request by AlCalzone**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: Hidden somewhere in https://github.com/ioBroker/ioBroker.iot/blob/master/lib/alexaSmartHomeV2.js, see linked issues for details.
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

fixes: https://github.com/ioBroker/adapter-core/issues/251
fixes: https://github.com/ioBroker/adapter-core/issues/252
fixes: https://github.com/ioBroker/adapter-core/issues/253

### CODE REVIEW

File: types/iobroker/index.d.ts [hunk 1]
```diff
@@ -138,6 +140,17 @@ declare global {
 
             /** Custom settings for this state */
             custom?: Record<string, any>;
+
+            /**
+             * Settings for IOT adapters and how the state should be named in e.g. Alexa.
+             * The string "ignore" is a special case, causing the state to be ignored.
+             */
+            smartName?: string | ({ [lang in Languages]?: string; } & {
```
Comment (UncleSamSwiss) [inline types/iobroker/index.d.ts:148]: Would it make sense to have an explicit `smartName?: 'ignore' | string | ...` (or is this not allowed?)

File: types/iobroker/index.d.ts [hunk 2]
```diff
@@ -170,7 +189,7 @@ declare global {
             enums?: Record<string, string>;
             type: string; // specified in the derived interfaces
             // Be strict with what we allow here. Read objects overwrite this with any.
-            common: StateCommon | ChannelCommon | DeviceCommon | OtherCommon;
+            common: StateCommon | ChannelCommon | DeviceCommon | FolderObject | EnumObject | OtherCommon;
```
Comment (UncleSamSwiss) [inline types/iobroker/index.d.ts:192]: Shouldn't it be `FolderCommon | EnumCommon`?

File: types/iobroker/index.d.ts [hunk 3]
```diff
@@ -61,6 +61,8 @@ declare global {
         type ObjectType = 'state' | 'channel' | 'device';
         type CommonType = 'number' | 'string' | 'boolean' | 'array' | 'object' | 'mixed' | 'file';
 
+        type Languages = "en" | "de" | "ru" | "pt" | "nl" | "fr" | "it" | "es" | "pl" | "zh-cn";
```
Comment (UncleSamSwiss) [inline types/iobroker/index.d.ts:64]: Single quote or double quote?

**Review by UncleSamSwiss — CHANGES_REQUESTED** [review]

File: types/iobroker/index.d.ts (hunk 1 above)
Comment (AlCalzone) [reply types/iobroker/index.d.ts:148]: `'ignore' | string` is identical to `string` (there is an open issue in the TS repo for auto completion purposes).

File: types/iobroker/index.d.ts (hunk 2 above)
Comment (AlCalzone) [reply types/iobroker/index.d.ts:192]: Oh, right!

**Review by UncleSamSwiss — APPROVED** [review]
Looks ok to me.

**Comment by AlCalzone** [comment]
Ready to merge

**Comment by AlCalzone** [comment]
@andrewbranch @orta it seems that the bot no longer informs us that the types are published.
In my [last PR](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45213) I saw the comment within a couple of hours.

**Pull request merged**